### PR TITLE
fix(sql): reduce memory utilization of cursor factories

### PR DESCRIPTION
**Pull request by bziobrowski**
This PR pushes most native memory allocations from factories to cursors .
Instead of being kept all the time (e.g. when factory is in cache), they're released on cursor.close()
and re-allocated on factory.getCursor(). 

Notes: 
1. some allocations can't be cleared completely on cursor.close() because it'd break the factory, e.g.  join metadata map  (moved to NATIVE_JOIN_MAP tag). 
2. NATIVE_OFFLOAD memory is ignored (for time being)
3. NATIVE_SAMPLE_BY_LONG_LIST is ignored because I couldn't make it grow visibly 
4. tags related to table reader and writer are ignored as well because they don't belong to cursor and are usually kept until timeout or pool closing .

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 193 / 202 (95.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|9|55.56%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashOuterJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryPages.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/analytic/RowNumberFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|28|28|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillPrevRecordCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinLightRecordCursorFactory.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 193 / 202 (95.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|9|55.56%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashOuterJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryPages.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/analytic/RowNumberFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|28|28|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillPrevRecordCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinLightRecordCursorFactory.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 194 / 202 (96.04%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|9|55.56%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashOuterJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryPages.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/analytic/RowNumberFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|28|28|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillPrevRecordCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinLightRecordCursorFactory.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/map/FastMap.java [hunk 1]
```diff
@@ -94,19 +105,33 @@ public FastMap(
             double loadFactor,
             HashFunction hashFunction,
             int maxResizes
+    ) {
+        this(pageSize, keyTypes, valueTypes, keyCapacity, loadFactor, hashFunction, maxResizes, -1);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/map/FastMap.java:109]: nit: we could pass actual tag we want and remove conditionals from the main constructor

File: core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java [hunk 2]
```diff
@@ -68,8 +66,8 @@ public LimitedSizeSortedLightRecordCursorFactory(
     @Override
     protected void _close() {
         base.close();
-        if (chain != null) {
-            chain.close();
+        if (cursor != null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java:69]: `Misc.free()`

File: core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java [hunk 3]
```diff
@@ -78,13 +76,12 @@ public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlEx
...
         try {
             initialize(executionContext, baseCursor);
-
             cursor.of(baseCursor, executionContext);
             return cursor;
         } catch (Throwable ex) {
             baseCursor.close();
-            if (chain != null) {
-                chain.clear();
+            if (cursor != null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java:83]: `Misc.free()`

File: core/src/main/java/io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java [hunk 4]
```diff
@@ -35,18 +35,27 @@ class SortedLightRecordCursor implements DelegatingRecordCursor {
...
 
     @Override
     public void close() {
-        chain.clear();
-        base.close();
+        if (isOpen) {
+            chain.close();
+            isOpen = false;
+        }
+        if (base != null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java:54]:
```
base = Misc.free(base);
baseRecord = null;
```

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 190 / 197 (96.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashOuterJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryPages.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/analytic/RowNumberFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|28|28|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillPrevRecordCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinLightRecordCursorFactory.java|2|2|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java [hunk 5]
```diff
@@ -80,6 +80,7 @@ public void setDouble(MapValue mapValue, double value) {
 
     @Override
     public void setNull(MapValue mapValue) {
+        mapValue.putDouble(valueIndex, Double.NaN);
```
[the remark this answers is not in the document]
Comment (amunra) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java:83]: Looks like this is a bugfix for `testSampleFillValueListWithNullAndPrev`. Nice.

**Review by amunra — APPROVED** [review]
Lookin' good.

The `Reallocatable` interface threw me off in terms of naming, but there's no reason why this can't be done as a small follow-up PR.

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 190 / 197 (96.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursorFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/HashOuterJoinRecordCursorFactory.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryPages.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/AvgDoubleGroupByFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/analytic/RowNumberFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|28|28|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursor.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillPrevRecordCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinLightRecordCursorFactory.java|2|2|100.00%|

**Pull request merged**